clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/cache/CatsOnDemandCacheUpdater.java

```diff
@@ -0,0 +1,224 @@
+package com.netflix.spinnaker.clouddriver.cache;
+
+import com.netflix.spinnaker.cats.agent.Agent;
+import com.netflix.spinnaker.cats.agent.AgentLock;
+import com.netflix.spinnaker.cats.agent.AgentScheduler;
+import com.netflix.spinnaker.cats.cache.CacheData;
+import com.netflix.spinnaker.cats.module.CatsModule;
+import com.netflix.spinnaker.cats.provider.Provider;
+import com.netflix.spinnaker.cats.provider.ProviderCache;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class CatsOnDemandCacheUpdater implements OnDemandCacheUpdater {
+
+  private static final Logger log = LoggerFactory.getLogger(CatsOnDemandCacheUpdater.class);
+
+  private final List<Provider> providers;
+  private final CatsModule catsModule;
+  // TODO(rz): Deliberately not using <? extends AgentLock> since it results in
+  //  compilation errors. This is a side-effect of migrating away from Groovy.
+  //  I'm sure there's a way, but it's the early morning and I'm pretty tired!
+  private final AgentScheduler agentScheduler;
```

Comment:
Since it's 3 hours later here...I think that in principle the way this would have to be solved is by parametrizing the class with `<T extends AgentLock>` and using `T` everywhere instead of `AgentLock`.  (The compiler is not smart enough to realize that the capture of `AgentLock` it got back from `agentScheduler.tryLock` is guaranteed to be of the same type it needs to pass back to `agentScheduler.lockValid` so we have to capture it ourselves.)

Reply:
Though maybe the real solution is to refactor `AgentScheduler` to make life easier for clients...seems like maybe what it really wants is to expose a `withLock(Closure<?>)` method or something like that; the lock type should be an implementation detail of the scheduler.

Reply:
That's a good call - perhaps I'll make that refactor separately, then work that into this PR.

Reply:
jk. I'll make that change as part of this PR - the methods are only used in this one class anyway.

Reply:
I lied. This refactor will impact a fair amount, so I'll knock it out separately after this PR.